File: ios/SuperWardrobe/SuperWardrobe/Services/SupabaseService.swift
import Foundation
import Supabase

@Observable
final class SupabaseService {
    static let shared = SupabaseService()

    let client: SupabaseClient

    private init() {
        client = SupabaseClient(
            supabaseURL: URL(string: Constants.supabaseURL)!,
            supabaseKey: Constants.supabaseAnonKey
        )
    }

    var currentUserId: UUID? {
        get async {
            try? await client.auth.session.user.id
        }
    }

    // MARK: - Auth

    func signIn(email: String, password: String) async throws {
        try await client.auth.signIn(email: email, password: password)
    }

    func signUp(email: String, password: String) async throws {
        try await client.auth.signUp(email: email, password: password)
    }

    func signOut() async throws {
        try await client.auth.signOut()
    }

    // MARK: - Clothing Items

    func fetchClothingItems(userId: UUID) async throws -> [ClothingItem] {
        try await client.from("clothing_items")
            .select()
            .eq("user_id", value: userId.uuidString)
            .eq("status", value: "active")
            .order("created_at", ascending: false)
            .execute()
            .value
    }

    func addClothingItem(_ item: ClothingItem) async throws {
        try await client.from("clothing_items")
            .insert(item)
            .execute()
    }

    func updateClothingItem(_ item: ClothingItem) async throws {
        try await client.from("clothing_items")
            .update(item)
            .eq("id", value: item.id.uuidString)
            .execute()
    }

    func deleteClothingItem(id: UUID) async throws {
        try await client.from("clothing_items")
            .update(["status": "deleted"])
            .eq("id", value: id.uuidString)
            .execute()
    }

    // MARK: - Categories

    func fetchCategories() async throws -> [Category] {
        try await client.from("categories")
            .select()
            .order("sort_order", ascending: true)
            .execute()
            .value
    }

    // MARK: - Outfits

    func fetchOutfits(userId: UUID) async throws -> [Outfit] {
        try await client.from("outfits")
            .select()
            .eq("user_id", value: userId.uuidString)
            .order("created_at", ascending: false)
            .execute()
            .value
    }

    func createOutfit(_ outfit: Outfit, items: [OutfitItem]) async throws {
        try await client.from("outfits")
            .insert(outfit)
            .execute()
        if !items.isEmpty {
            try await client.from("outfit_items")
                .insert(items)
                .execute()
        }
    }

    // MARK: - Outfit Diary

    func fetchOutfitDiary(userId: UUID, month: Date) async throws -> [OutfitDiary] {
        let calendar = Calendar.current
        let startOfMonth = calendar.date(from: calendar.dateComponents([.year, .month], from: month))!
        let endOfMonth = calendar.date(byAdding: .month, value: 1, to: startOfMonth)!

        let formatter = ISO8601DateFormatter()
        return try await client.from("outfit_diary")
            .select()
            .eq("user_id", value: userId.uuidString)
            .gte("date", value: formatter.string(from: startOfMonth))
            .lt("date", value: formatter.string(from: endOfMonth))
            .order("date", ascending: false)
            .execute()
            .value
    }

    func saveOutfitDiary(_ diary: OutfitDiary) async throws {
        try await client.from("outfit_diary")
            .upsert(diary)
            .execute()
    }

    // MARK: - Daily Recommendations

    func fetchDailyRecommendations(userId: UUID, date: Date) async throws -> [DailyRecommendation] {
        let formatter = ISO8601DateFormatter()
        return try await client.from("daily_recommendations")
            .select()
            .eq("user_id", value: userId.uuidString)
            .eq("date", value: formatter.string(from: date))
            .execute()
            .value
    }

    // MARK: - Purchase Recommendations

    func fetchPurchaseRecommendations(userId: UUID) async throws -> [PurchaseRecommendation] {
        try await client.from("purchase_recommendations")
            .select()
            .eq("user_id", value: userId.uuidString)
            .eq("status", value: "active")
            .order("priority", ascending: true)
            .execute()
            .value
    }

    // MARK: - Travel Plans

    func fetchTravelPlans(userId: UUID) async throws -> [TravelPlan] {
        try await client.from("travel_plans")
            .select()
            .eq("user_id", value: userId.uuidString)
            .order("start_date", ascending: true)
            .execute()
            .value
    }

    func createTravelPlan(_ plan: TravelPlan) async throws {
        try await client.from("travel_plans")
            .insert(plan)
            .execute()
    }

    func addTravelPlanOutfit(planId: UUID, date: Date, outfitId: UUID) async throws {
        let item = TravelPlanOutfit(
            id: UUID(),
            travelPlanId: planId,
            date: date,
            outfitId: outfitId
        )
        try await client.from("travel_plan_outfits")
            .insert(item)
            .execute()
    }

    // MARK: - User Profile

    func fetchUserProfile(userId: UUID) async throws -> UserProfile? {
        let profiles: [UserProfile] = try await client.from("user_profiles")
            .select()
            .eq("id", value: userId.uuidString)
            .execute()
            .value
        return profiles.first
    }

    func updateUserProfile(_ profile: UserProfile) async throws {
        try await client.from("user_profiles")
            .upsert(profile)
            .execute()
    }

    // MARK: - Storage

    func uploadImage(data: Data, bucket: String, path: String) async throws -> URL {
        try await client.storage.from(bucket).upload(
            path: path,
            file: data,
            options: FileOptions(contentType: "image/jpeg")
        )
        let publicURL = try client.storage.from(bucket).getPublicURL(path: path)
        return publicURL
    }
}

File: android/superwardrobe/app/src/main/java/com/superwardrobe/service/SupabaseService.kt
package com.superwardrobe.service

import com.superwardrobe.data.model.*
import com.superwardrobe.util.Constants
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.Auth
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.auth.providers.builtin.Email
import io.github.jan.supabase.createSupabaseClient
import io.github.jan.supabase.postgrest.Postgrest
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import io.github.jan.supabase.postgrest.query.Order
import io.github.jan.supabase.storage.Storage
import io.github.jan.supabase.storage.storage
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

object SupabaseService {

    val client: SupabaseClient = createSupabaseClient(
        supabaseUrl = Constants.SUPABASE_URL,
        supabaseKey = Constants.SUPABASE_ANON_KEY
    ) {
        install(Auth)
        install(Postgrest)
        install(Storage)
    }

    suspend fun signIn(email: String, password: String) = withContext(Dispatchers.IO) {
        client.auth.signInWith(Email) {
            this.email = email
            this.password = password
        }
    }

    suspend fun signUp(email: String, password: String) = withContext(Dispatchers.IO) {
        client.auth.signUpWith(Email) {
            this.email = email
            this.password = password
        }
    }

    suspend fun signOut() = withContext(Dispatchers.IO) {
        client.auth.signOut()
    }

    fun getCurrentUserId(): String? {
        return client.auth.currentUserOrNull()?.id
    }

    // ClothingItem CRUD
    suspend fun getClothingItems(userId: String): List<ClothingItem> = withContext(Dispatchers.IO) {
        client.from("clothing_items")
            .select {
                filter { eq("user_id", userId) }
                order("created_at", Order.DESCENDING)
            }
            .decodeList()
    }

    suspend fun getClothingItemById(id: String): ClothingItem = withContext(Dispatchers.IO) {
        client.from("clothing_items")
            .select {
                filter { eq("id", id) }
            }
            .decodeSingle()
    }

    suspend fun insertClothingItem(item: ClothingItem): ClothingItem = withContext(Dispatchers.IO) {
        client.from("clothing_items")
            .insert(item) { select() }
            .decodeSingle()
    }

    suspend fun updateClothingItem(item: ClothingItem) = withContext(Dispatchers.IO) {
        client.from("clothing_items")
            .update(item) {
                filter { eq("id", item.id) }
            }
    }

    suspend fun deleteClothingItem(id: String) = withContext(Dispatchers.IO) {
        client.from("clothing_items")
            .delete {
                filter { eq("id", id) }
            }
    }

    suspend fun searchClothingItems(userId: String, query: String): List<ClothingItem> = withContext(Dispatchers.IO) {
        client.from("clothing_items")
            .select {
                filter {
                    eq("user_id", userId)
                    or {
                        ilike("name", "%$query%")
                        ilike("brand", "%$query%")
                        ilike("color", "%$query%")
                    }
                }
            }
            .decodeList()
    }

    // Categories
    suspend fun getCategories(): List<Category> = withContext(Dispatchers.IO) {
        client.from("categories")
            .select {
                order("sort_order", Order.ASCENDING)
            }
            .decodeList()
    }

    // Outfits
    suspend fun getOutfits(userId: String): List<Outfit> = withContext(Dispatchers.IO) {
        client.from("outfits")
            .select {
                filter { eq("user_id", userId) }
                order("created_at", Order.DESCENDING)
            }
            .decodeList()
    }

    suspend fun insertOutfit(outfit: Outfit): Outfit = withContext(Dispatchers.IO) {
        client.from("outfits")
            .insert(outfit) { select() }
            .decodeSingle()
    }

    suspend fun updateOutfit(outfit: Outfit) = withContext(Dispatchers.IO) {
        client.from("outfits")
            .update(outfit) {
                filter { eq("id", outfit.id) }
            }
    }

    suspend fun deleteOutfit(id: String) = withContext(Dispatchers.IO) {
        client.from("outfits")
            .delete {
                filter { eq("id", id) }
            }
    }

    // OutfitDiary
    suspend fun getOutfitDiaries(userId: String): List<OutfitDiary> = withContext(Dispatchers.IO) {
        client.from("outfit_diaries")
            .select {
                filter { eq("user_id", userId) }
                order("date", Order.DESCENDING)
            }
            .decodeList()
    }

    suspend fun insertOutfitDiary(diary: OutfitDiary): OutfitDiary = withContext(Dispatchers.IO) {
        client.from("outfit_diaries")
            .insert(diary) { select() }
            .decodeSingle()
    }

    suspend fun updateOutfitDiary(diary: OutfitDiary) = withContext(Dispatchers.IO) {
        client.from("outfit_diaries")
            .update(diary) {
                filter { eq("id", diary.id) }
            }
    }

    // DailyRecommendation
    suspend fun getDailyRecommendations(userId: String, date: String): List<DailyRecommendation> = withContext(Dispatchers.IO) {
        client.from("daily_recommendations")
            .select {
                filter {
                    eq("user_id", userId)
                    eq("date", date)
                }
            }
            .decodeList()
    }

    suspend fun insertDailyRecommendation(rec: DailyRecommendation): DailyRecommendation = withContext(Dispatchers.IO) {
        client.from("daily_recommendations")
            .insert(rec) { select() }
            .decodeSingle()
    }

    suspend fun markRecommendationWorn(id: String) = withContext(Dispatchers.IO) {
        client.from("daily_recommendations")
            .update(mapOf("is_worn" to true)) {
                filter { eq("id", id) }
            }
    }

    // PurchaseRecommendation
    suspend fun getPurchaseRecommendations(userId: String): List<PurchaseRecommendation> = withContext(Dispatchers.IO) {
        client.from("purchase_recommendations")
            .select {
                filter { eq("user_id", userId) }
                order("priority", Order.DESCENDING)
            }
            .decodeList()
    }

    // TravelPlan
    suspend fun getTravelPlans(userId: String): List<TravelPlan> = withContext(Dispatchers.IO) {
        client.from("travel_plans")
            .select {
                filter { eq("user_id", userId) }
                order("start_date", Order.ASCENDING)
            }
            .decodeList()
    }

    suspend fun insertTravelPlan(plan: TravelPlan): TravelPlan = withContext(Dispatchers.IO) {
        client.from("travel_plans")
            .insert(plan) { select() }
            .decodeSingle()
    }

    suspend fun updateTravelPlan(plan: TravelPlan) = withContext(Dispatchers.IO) {
        client.from("travel_plans")
            .update(plan) {
                filter { eq("id", plan.id) }
            }
    }

    suspend fun deleteTravelPlan(id: String) = withContext(Dispatchers.IO) {
        client.from("travel_plans")
            .delete {
                filter { eq("id", id) }
            }
    }

    // UserProfile
    suspend fun getUserProfile(userId: String): UserProfile? = withContext(Dispatchers.IO) {
        client.from("user_profiles")
            .select {
                filter { eq("id", userId) }
            }
            .decodeList<UserProfile>()
            .firstOrNull()
    }

    suspend fun upsertUserProfile(profile: UserProfile) = withContext(Dispatchers.IO) {
        client.from("user_profiles")
            .upsert(profile)
    }

    // Storage
    suspend fun uploadImage(bucket: String, path: String, data: ByteArray): String = withContext(Dispatchers.IO) {
        client.storage.from(bucket).upload(path, data)
        client.storage.from(bucket).publicUrl(path)
    }
}
